src/testing/advection_limiter.cpp

```diff
@@ -0,0 +1,231 @@
+                                }
+                                l2error += pow(soln_at_q[istate] - uexact, 2) * fe_values_extra.JxW(iquad);
+                            }
+                        }
+
+                    }
+                    const double l2error_mpi_sum = std::sqrt(dealii::Utilities::MPI::sum(l2error, this->mpi_communicator));
+
+                    // Convergence table
+                    const double dx = 1.0 / pow(n_dofs, (1.0 / dim));
+                    grid_size[igrid] = dx;
+                    soln_error[igrid] = l2error_mpi_sum;
+
+                    convergence_table.add_value("p", poly_degree);
+                    convergence_table.add_value("cells", n_global_active_cells);
+                    convergence_table.add_value("DoFs", n_dofs);
+                    convergence_table.add_value("dx", dx);
+                    convergence_table.add_value("soln_L2_error", l2error_mpi_sum);
+
+                    this->pcout << " Grid size h: " << dx
+                        << " L2-soln_error: " << l2error_mpi_sum
+                        << " Residual: " << ode_solver->residual_norm
+                        << std::endl;
+
+                    if (igrid > igrid_start) {
+                        const double slope_soln_err = log(soln_error[igrid] / soln_error[igrid - 1])
+                            / log(grid_size[igrid] / grid_size[igrid - 1]);
+                        this->pcout << "From grid " << igrid - 1
+                            << "  to grid " << igrid
+                            << "  dimension: " << dim
+                            << "  polynomial degree p: " << poly_degree
+                            << std::endl
+                            << "  solution_error1 " << soln_error[igrid - 1]
+                            << "  solution_error2 " << soln_error[igrid]
+                            << "  slope " << slope_soln_err
+                            << std::endl;
+                    }
+
+                    this->pcout << " ********************************************"
+                        << std::endl
+                        << " Convergence rates for p = " << poly_degree
+                        << std::endl
+                        << " ********************************************"
+                        << std::endl;
+                    convergence_table.evaluate_convergence_rates("soln_L2_error", "cells", dealii::ConvergenceTable::reduction_rate_log2, dim);
+                    convergence_table.set_scientific("dx", true);
+                    convergence_table.set_scientific("soln_L2_error", true);
+                    if (this->pcout.is_active()) convergence_table.write_text(this->pcout.get_stream());
+                }
+            }//end of grid loop
+            //want to add some condition to check for if(){return 1} else{return 0}
+            return 0; //if got to here means passed the test, otherwise would've failed earlier
```

Comment:
Is there any condition you could check rather than returning `0` always? For instance, can you check that the limiters are behaving as expected by checking convergence rate, or by checking that positivity/maximum principle is indeed satisfied?

Reply:
I left the return condition as 0 since the limiters have checks in them for whether they satisfy the principle they're designed to preserve. I could potentially do a different check for the convergence tests but I won't be able to set up a generic condition such as `soln_L2_error > poly_degree+1` since the accuracy is expected to degenerate for certain cases.

Reply:
Okay, that sounds reasonable. In general, you should try to have a pass/fail test in the test itself, but if the limiter will catch any weird behaviour, it's fine by me to leave as-is.

You should remove the line `//want to add some condition to check for if(){return 1} else{return 0}` from this file and the other testing files if there's no suitable test!

Reply:
With limiters, is there an expected criterion that we expect to occur after limiting a solution? Like, if we run the flow solver with limiters, can we somehow check that density is non-negative everywhere (including inbetween collocated solution points)? Or that there is some kind of TVB/TVD property between the inital and final solution?

Reply:
I implemented checks for the cases within flow solver to verify that the principle for each limiter is preserved! :smile: I don't have it checking between collocated solution points however, it just checks at the solution points.

Reply:
@cpethrick can we resolve this?

Reply:
Yep, the check in flow solver case looks okay.